File: usvm-ts/src/test/resources/samples/lang/Cast.ts
// @ts-nocheck
// noinspection JSUnusedGlobalSymbols

class Cast {
    castAnyToNumber(x: any): number {
        const y = x as number;
        if (y === 42) {
            return 1;
        }
        return 0;
    }

    castAnyToBoolean(x: any): number {
        const y = x as boolean;
        if (y === true) {
            return 1;
        }
        if (y === false) {
            return 2;
        }
        return -1;
    }

    castAnyToString(x: any): number {
        const y = x as string;
        if (y === "test") {
            return 1;
        }
        return 0;
    }

    castNumberToAny(x: number): number {
        const y = x as any;
        if (y === 42) {
            return 1;
        }
        return 0;
    }

    castBooleanToAny(x: boolean): number {
        const y = x as any;
        if (y === true) {
            return 1;
        }
        if (y === false) {
            return 2;
        }
        return -1; // unreachable
    }

    castWithMultipleBranches(x: any): number {
        const y = x as number;
        if (y > 10) {
            return 1;
        } else if (y > 5) {
            return 2;
        } else if (y > 0) {
            return 3;
        }
        return 0;
    }

    castObjectToInterface(obj: any): number {
        const typed = obj as { value: number };
        if (typed.value === 100) {
            return 100;
        }
        return 0;
    }

    castNullableToNumber(x: number | null): number {
        if (x === null) {
            return -1;
        }
        const y = x as number;
        if (y === 5) {
            return 5;
        }
        return 0;
    }

    castChained(x: any): number {
        const y = x as any;
        const z = y as number;
        if (z === 7) {
            return 7;
        }
        return 0;
    }

    castInExpression(x: any): number {
        if ((x as number) === 3) {
            return 3;
        }
        return 0;
    }

    castAndArithmetic(x: any): number {
        const y = x as number;
        const result = y + 10;
        if (result === 20) {
            return 1;
        }
        return 0;
    }

    castUnionType(x: number | string): number {
        const y = x as number;
        if (y === 15) {
            return 15;
        }
        return 0;
    }
}

File: usvm-ts/src/test/kotlin/org/usvm/samples/lang/Cast.kt
package org.usvm.samples.lang

import org.jacodb.ets.model.EtsScene
import org.junit.jupiter.api.Disabled
import org.usvm.api.TsTestValue
import org.usvm.util.TsMethodTestRunner
import org.usvm.util.eq
import org.usvm.util.neq
import kotlin.test.Test

class Cast : TsMethodTestRunner() {
    private val tsPath = "/samples/lang/Cast.ts"

    override val scene: EtsScene = loadScene(tsPath)

    @Test
    fun `cast any to number`() {
        val method = getMethod("castAnyToNumber")
        discoverProperties<TsTestValue, TsTestValue.TsNumber>(
            method = method,
            { _, r -> r eq 0 },
            { _, r -> r eq 1 },
        )
    }

    @Test
    fun `cast any to boolean`() {
        val method = getMethod("castAnyToBoolean")
        discoverProperties<TsTestValue.TsBoolean, TsTestValue.TsNumber>(
            method = method,
            { x, r -> x.value && (r eq 1) },
            { x, r -> !x.value && (r eq 2) },
            invariants = arrayOf(
                { x, r ->
                    if (x.value) {
                        r eq 1
                    } else {
                        r eq 2
                    }
                }
            )
        )
    }

    @Disabled("Input strings are not supported")
    @Test
    fun `cast any to string`() {
        val method = getMethod("castAnyToString")
        discoverProperties<TsTestValue, TsTestValue.TsNumber>(
            method = method,
            { _, r -> r eq 1 },
            { _, r -> r eq 0 },
        )
    }

    @Test
    fun `cast number to any`() {
        val method = getMethod("castNumberToAny")
        discoverProperties<TsTestValue.TsNumber, TsTestValue.TsNumber>(
            method = method,
            { x, r -> (x eq 42) && (r eq 1) },
            { x, r -> (x neq 42) && (r eq 0) },
        )
    }

    @Test
    fun `cast boolean to any`() {
        val method = getMethod("castBooleanToAny")
        discoverProperties<TsTestValue.TsBoolean, TsTestValue.TsNumber>(
            method = method,
            { x, r -> x.value && (r eq 1) },
            { x, r -> !x.value && (r eq 2) },
            invariants = arrayOf(
                { x, r ->
                    if (x.value) {
                        r eq 1
                    } else {
                        r eq 2
                    }
                }
            )
        )
    }

    @Test
    fun `cast with multiple branches`() {
        val method = getMethod("castWithMultipleBranches")
        discoverProperties<TsTestValue, TsTestValue.TsNumber>(
            method = method,
            { x, r -> (r eq 0) && (x is TsTestValue.TsNumber) && (x.number <= 0) },
            { x, r -> (r eq 1) && (x is TsTestValue.TsNumber) && (x.number > 10) },
            { x, r -> (r eq 2) && (x is TsTestValue.TsNumber) && (x.number <= 10) && (x.number > 5) },
            { x, r -> (r eq 3) && (x is TsTestValue.TsNumber) && (x.number <= 5) && (x.number > 0) },
        )
    }

    @Test
    fun `cast object to interface`() {
        val method = getMethod("castObjectToInterface")
        discoverProperties<TsTestValue, TsTestValue.TsNumber>(
            method = method,
            { _, r -> r eq 100 },
            { _, r -> r eq 0 },
        )
    }

    @Test
    fun `cast nullable to number`() {
        val method = getMethod("castNullableToNumber")
        discoverProperties<TsTestValue, TsTestValue.TsNumber>(
            method = method,
            { x, r -> r eq -1 },
            { x, r -> r eq 5 },
            { x, r -> r eq 0 },
        )
    }

    @Test
    fun `cast chained`() {
        val method = getMethod("castChained")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r eq 7 },
            { r -> r eq 0 },
        )
    }

    @Test
    fun `cast in expression`() {
        val method = getMethod("castInExpression")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r eq 3 },
            { r -> r eq 0 },
        )
    }

    @Test
    fun `cast and arithmetic`() {
        val method = getMethod("castAndArithmetic")
        discoverProperties<TsTestValue, TsTestValue.TsNumber>(
            method = method,
            { x, r -> r eq 1 },
            { x, r -> r eq 0 },
        )
    }

    @Test
    fun `cast union type`() {
        val method = getMethod("castUnionType")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r eq 15 },
            { r -> r eq 0 },
        )
    }
}
